### PR TITLE
Port twisted.protocols.sip to Python 3

### PR DESCRIPTION
**Pull request by rodrigc**
https://twistedmatrix.com/trac/ticket/8669

### CODE REVIEW

File: src/twisted/protocols/sip.py [hunk 1]
```diff
@@ -109,9 +110,13 @@
 
 
 def dashCapitalize(s):
-    ''' Capitalize a string, making sure to treat - as a word separator '''
+    '''
+    Capitalize a string, making sure to treat - as a word separator
```
Comment (moshez) [inline src/twisted/protocols/sip.py]: This is hard to read. Maybe "to treat '-' as a word separator"?

File: src/twisted/protocols/sip.py [hunk 2]
```diff
@@ -543,59 +583,75 @@ class MessagesParser(basic.LineReceiver):
...
             self.reset()
         else:
-            # we have enough data and message wasn't finished? something is wrong
+            # We have enough data and message wasn't finished? something is wrong
             raise RuntimeError("this should never happen")
 
+
     def dataReceived(self, data):
         try:
+            if isinstance(data, unicode):
```
Comment (moshez) [inline src/twisted/protocols/sip.py:632]: Why is this needed? Can we fix the calling site instead?

File: src/twisted/protocols/sip.py [hunk 3]
```diff
@@ -543,59 +583,75 @@ class MessagesParser(basic.LineReceiver):
...
         raise NotImplementedError
 
+
     def lineLengthExceeded(self, line):
         self.invalidMessage()
 
+
     def lineReceived(self, line):
+        if _PY3 and isinstance(line, bytes):
+            line = line.decode("utf-8")
```
Comment (moshez) [inline src/twisted/protocols/sip.py:653]: This seems like there should be an abstraction "force_bytes" which will do it?

File: src/twisted/protocols/sip.py [hunk 4]
```diff
@@ -754,76 +828,104 @@ def sendMessage(self, destURL, message):
             raise RuntimeError("only UDP currently supported")
         if self.debug:
             log.msg("Sending %r to %r" % (message.toString(), destURL))
-        self.transport.write(message.toString(), (destURL.host, destURL.port or self.PORT))
+        data = message.toString()
+        if isinstance(data, unicode):
+            data = data.encode("utf-8")
```
Comment (moshez) [inline src/twisted/protocols/sip.py:833]: Is it OK to assume message.toString() will always return a unicode string that we can encode into utf-8? The doc string doesn't specify what assumptions it makes on message

**Review by moshez — APPROVED** [review]
Some small comments. Feel free to address those you think are good and push.

File: src/twisted/protocols/sip.py (hunk 2 above)
Comment (rodrigc) [reply src/twisted/protocols/sip.py:632]: Clients outside of Twisted can be written with things like dataReceived("someString")
so that means in Py2 it is bytes, but in Py3 it is unicode.  So we need this conversion.
